// Showcasing the capabilities of protobuf with gRPC

syntax = "proto3";

import "google/protobuf/timestamp.proto";

option go_package = "/operations_ecosys";

package operations_ecosys;

// Note for updates:
// If the object fields should be updated, 
// fields that do not need to be updated can be blank.
// However, do note that enums and bools have default values, 
// hence enums and bools should always be filled. 

service AdminServices {
    // User
    rpc AddUser(User) returns (Response) {}
    rpc UpdateUser(User) returns (Response) {}
    rpc DeleteUser(User) returns (Response) {}
    // TODO change user response to have user scoring and stuff
    rpc FindUsers(UserQuery) returns (stream UsersResponse) {}

    // Client
    rpc AddClient(Client) returns (Response) {}
    rpc UpdateClient(Client) returns (Response) {}
    rpc DeleteClient(Client) returns (Response) {}
    rpc FindClients(ClientQuery) returns (stream ClientResponse) {}
}

service BroadcastServices{
    rpc AddBroadcast(Broadcast) returns (Response) {}
    // Note that this update does not update the broadcast's recipient's inner status 
    // such as the acknowledgement or rejection status but only if the recipient 
    // is part of the broadcast.
    rpc UpdateBroadcast(Broadcast) returns (Response) {}
    rpc DeleteBroadcast(Broadcast) returns (Response) {}
    rpc FindBroadcasts(BroadcastQuery) returns (stream BroadcastResponse) {}

    // Updating of broadcast recipients
    rpc UpdateBroadcastRecipient(BroadcastRecipient) returns (Response) {}
}

service RosterServices{
    // Add multiple rosters for different AIFS at the same time
    rpc AddRoster(BulkRosters) returns (Response) {}
    // Note that this update does not update the roster's guard's inner status 
    // such as the acknowledgement or attended status but only if the guard 
    // is part of the roster.
    rpc UpdateRoster(Roster) returns (Response) {}
    rpc DeleteRoster(Roster) returns (Response) {}
    rpc FindRosters(RosterQuery) returns (stream RosterResponse) {}
    rpc GetAvailableUsers(AvailabilityQuery) returns (stream EmployeeEvaluationResponse) {}
}

// Admin
/***************************************************************************
*                              USER MESSAGES                               *
****************************************************************************/
// All users who use the operations ecosystem.
// TODO: decide if we want to have an inactive field for users
//       so that we don't need to delete users completely. 
message User {
    int64 user_id = 1;
    UserType user_type = 2;
    string name = 3;
    string email = 4;
    string phone_number = 5;
    string telegram_handle = 6;
    string user_security_img = 7;
    bool is_part_timer = 8;
    int64 tele_chat_id = 9;

    enum UserType {
        ISPECIALIST = 0;
        SECURITY_GUARD = 1;
        CONTROLLER = 2;
        MANAGER = 3;
    }
}

// Passing around multiple users in one message. 
message UsersResponse {
    Response response = 1;
    User user = 2;
}

// Filter the types of users to be returned. 
message UserFilter {
    Field field = 1;
    Filter comparisons = 2;

    // More fields can be added in the future. 
    enum Field{
        USER_ID = 0;
        TYPE = 1;
        NAME = 2;
        EMAIL = 3;
        PHONE_NUMBER = 4;
        TELEGRAM_HANDLE = 5;
        IS_PART_TIMER = 6;
    }
}

// Get specific types users as specified in the Filter. 
// If one wants to get all objects, leave filters empty. 
// A default limit of 10 will be used if the field is empty. 
message UserQuery {
    repeated UserFilter filters = 1;
    // Limit the number of objects being returned. 
    // If only 5 objects should be shown, limit = 5;
    int64 limit = 2;

    // Skip n rows from the database
    int64 skip = 3;
    
    // Order the queries, by default the order is desc by creation date
    OrderByUser order_by = 4;
}


message OrderByUser{
    UserFilter.Field field = 1;
    OrderBy order_by = 2;
}

/***************************************************************************
*                              CLIENT MESSAGES                              *
****************************************************************************/

// All clients who hire the company to protect their compounds. 
// TODO: decide if we want to have an inactive field for clients
//       so that we don't need to delete cleints completely. 
message Client {
    int64 client_id = 1;
    string name = 2;
    string abbreviation = 3;
    string email = 4;
    string address = 5;
    int64 postal_code = 6;
    string phone_number = 7;
}

message ClientResponse {
    Response response = 1;
    Client client = 2;
}

// Filter the types of clients to be returned. 
message ClientFilter {
    Field field = 1;
    Filter comparisons = 2;

    // More fields can be added in the future. 
    enum Field{
        CLIENT_ID = 0;
    }
}

// Get specific types clients as specified in the Filter. 
// If one wants to get all objects, leave filters empty. 
// A default limit of 10 will be used if the field is empty. 
message ClientQuery {
    repeated ClientFilter filters = 1;
    // Limit the number of objects being returned. 
    // If only 5 objects should be shown, limit = 5;
    int64 limit = 2;

    // Skip n rows from the database
    int64 skip = 3;
    
    // Order the queries, by default the order is desc by creation date
    OrderByClient order_by = 4;
}


message OrderByClient{
    ClientFilter.Field field = 1;
    OrderBy order_by = 2;
}

/***************************************************************************
*                          BROADCASTING MESSAGES                           *
****************************************************************************/
// The default fields of a broadcast
message Broadcast {
    // Broadcast IDs are only useful for the backend database. 
    int64 broadcast_id = 1;
    BroadcastType type = 2;
    string content = 3;
    google.protobuf.Timestamp creation_date = 4;
    google.protobuf.Timestamp deadline = 5;
    // The whole user fields does not need to be filled, 
    // as long as the user is identifiable.
    User creator = 6;
    repeated AIFSBroadcastRecipient recipients = 7;
    UrgencyType urgency = 8;
    
    enum BroadcastType{
        ANNOUNCEMENT = 0;
        ASSIGNMENT = 1;
    }

    enum UrgencyType{
        LOW = 0;
        MEDIUM = 1;
        HIGH = 2;
    }
}

message AIFSBroadcastRecipient {
    repeated BroadcastRecipient recipient =1;
    int64 aifs_id = 2;
}

message BroadcastRecipient {
    int64 broadcast_recipients_id = 1;
    User recipient = 2;
    bool acknowledged = 3;
    bool rejected = 4;
    google.protobuf.Timestamp last_replied = 5;
    int64 aifs_id = 6;
}


// Passing around multiple broadcasts in one message. 
message BroadcastResponse {
    Response response = 1;
    Broadcast broadcast = 2;
}

// Filter the types of broadcasts to be returned. 
message BroadcastFilter {
    Field field = 1;
    Filter comparisons = 2;

    // More fields can be added in the future. 
    enum Field{
        BROADCAST_ID = 0;
        TYPE = 1;
        CONTENT = 2;
        CREATION_DATE = 3;
        DEADLINE = 4;
        // Creator and recipient values should 
        // be the user id of these users
        CREATOR_ID = 5;
        // Note: Single recipient
        RECEIPEIENT_ID = 6;
        NUM_RECEIPIENTS = 7;
        URGENCY = 8;
        AIFS_ID = 9;
        BROADCAST_RECIPIENT_TABLE_ID = 10;
    }
}

// Get specific types users as specified in the Filter. 
// If one wants to get all objects, leave filters empty. 
// A default limit of 10 will be used if the field is empty. 
message BroadcastQuery {
    repeated BroadcastFilter filters = 1;
    // Limit the number of objects being returned. 
    // If only 5 objects should be shown, limit = 5;
    int64 limit = 2; 
    // Skip n rows from the database
    int64 skip = 3;

    // Order the queries, by default the order is desc by creation date
    OrderByBroadcast order_by = 4;
}

message OrderByBroadcast{
    BroadcastFilter.Field field = 1;
    OrderBy order_by = 2;
}


/***************************************************************************
*                            ROSTERING MESSAGES                            *
****************************************************************************/
// The default fields of a roster
message Roster {
    // Roster IDs are only useful for the backend database. 
    int64 rostering_id = 1;
    int64 aifs_id = 2;
    google.protobuf.Timestamp start_time = 3;
    google.protobuf.Timestamp end_time = 4;
    repeated AIFSClientRoster clients = 5;
    repeated RosterAssignement guard_assigned = 6;

    // The roster is not actually in the database
    bool is_default = 7;
    Status status = 8;

    enum Status {
        PENDING = 0;
        CONFIRMED = 1;
        REJECTED = 2;
    }
}

message AIFSClientRoster {
    int64 aifs_client_roster_id = 1;
    Client client = 2;    
    int64 patrol_order = 3;
}

message RosterAssignement {
    int64 roster_assignment_id = 1;
    // The whole user fields does not need to be filled, 
    // as long as the user is identifiable.
    EmployeeEvaluation guard_assigned = 2;
    google.protobuf.Timestamp custom_start_time = 3;
    google.protobuf.Timestamp custom_end_time = 4;
    bool confirmed = 5;
    bool attended = 6;
    google.protobuf.Timestamp attendance_time = 7;
    // If the assignment is part of the current assignment
    // or if it was previously assigned and is now removed
    bool is_assigned = 8;
    bool rejected = 9;
}

message BulkRosters{
    repeated Roster rosters = 1;
}

// Passing around multiple broadcasts in one message. 
message RosterResponse {
    Response response = 1;
    Roster roster = 2;
}

// Filter the types of rosters to be returned. 
message RosterFilter {
    Field field = 1;
    Filter comparisons = 2;

    // More fields can be added in the future. 
    enum Field{
        ROSTER_ID = 0;
        ROSTER_ASSIGNMENT_ID = 1;
        ROSTER_AIFS_CLIENT_ID = 2;
        AIFS_ID = 3;
        // Note: single user ID
        GUARD_ASSIGNED_ID = 4;
        // Note: single client ID
        CLIENT_ID = 5;
        GUARD_ASSIGNMENT_CONFIRMATION = 6;
        GUARD_ASSIGNMENT_ATTENDED = 7;
        START_TIME = 8;
        END_TIME = 9;
        IS_ASSIGNED = 10;
        DEFAULT_ROSTERING_DAY_OF_WEEK = 11;
        GUARD_ASSIGNMENT_REJECTION = 12;
    }
}

// Get specific types rosters as specified in the Filter. 
// If one wants to get all objects, leave filters empty. 
// A default limit of 10 will be used if the field is empty. 
message RosterQuery {
    repeated RosterFilter filters = 1;
    // Limit the number of objects being returned. 
    // If only 5 objects should be shown, limit = 5;
    int64 limit = 2; 
    // Skip n rows from the database
    int64 skip = 3;

    // Order the queries, by default the order is desc by creation date
    OrderByRoster order_by = 4;
}

message OrderByRoster{
    RosterFilter.Field field = 1;
    OrderBy order_by = 2;
}

/***************************************************************************
*                 ROSTERING AVAILABILITY MESSAGES                          *
****************************************************************************/

// Note that this query only checks for users 
// available for the whole time between a start 
// and end time.

// TODO: Find out if the user should be available
// throughout the whole duration. 
message AvailabilityQuery {
    string start_time = 1;
    // If end time is null, the time period is taken 
    // to be from the start time + 12h
    string end_time = 2;

    // Limit the number of objects being returned. 
    // If only 5 objects should be shown, limit = 5;
    int64 limit = 3; 
    // Skip n rows from the database
    int64 skip = 4;

    // The client has no current need to fill this up.
    repeated AvailabilityFilter filters = 5;
    OrderByQuery order_by = 6;
}

message OrderByQuery{
    AvailabilityFilter.Field field = 1;
    OrderBy order_by = 2;
}



// Filter the types of availabilty to be returned. 
message AvailabilityFilter {
    Field field = 1;
    Filter comparisons = 2;

    // More fields can be added in the future. 
    enum Field{
        AVAILABILITY_ID = 0;
        // Week of the year
        WEEK = 1;
        YEAR = 2;
        // Note: single user ID
        GUARD_ID = 3;

        // Day of the week to look at
        // The values for these fields in Filter comparisons
        // should be an empty string
        SUN = 4;
        MON = 5;
        TUES = 6;
        WED = 7;
        THURS = 8;
        FRI = 9;
        SAT = 10;
        NEXT_SUN = 11;
    }
}

/***************************************************************************
*                ROSTERING EMPLOYEE EVALUATION MESSAGES                    *
****************************************************************************/
message EmployeeEvaluationResponse {
    Response response = 1;
    EmployeeEvaluation employee = 2;
}

message EmployeeEvaluation {
    User employee = 1;
    float employee_score = 2;

    // This is used if it is needed to determine 
    // if the employee is available within a previously 
    // specified time.
    bool is_available = 3; 
}

/***************************************************************************
*                              COMMON MESSAGES                              *
****************************************************************************/

// Generic reponses to add or update requests
message Response {
    Type type = 1;
    string error_message = 2;
    // Return any pk of the row that the query modified
    int64 primary_key = 3;

    enum Type {
        ACK = 0;
        ERROR = 1;
    }
}

// This is used to indicate what kind of objects should be returned that
// fit this critera. 
// For example, if one wishes to get all broadcasts that have more than 
// one recipient. They might put the comparison as GREATER, value = 1;
// The field to be compared with is in the corresponding XXXFilter message.
message Filter {
    Comparisons comparison = 1;
    string value = 2;

    enum Comparisons{
        GREATER = 0;
        GREATER_EQ = 1;
        EQUAL = 2;
        LESSER_EQ = 3;
        LESSER = 4;
        CONTAINS = 5;
        IN = 6;
        NOT_IN = 7;
    }
}

enum OrderBy {
    ASC = 0;
    DESC = 1;
}
